crates/schema-inference/src/server.rs

```diff
@@ -0,0 +1,277 @@
+use std::time::Duration;
+
+use crate::inference::infer_shape;
+use crate::json_decoder::JsonCodec;
+use crate::schema::SchemaBuilder;
+use crate::shape;
+
+use assemble::journal_selector;
+use doc::inference::Shape;
+use futures::{Stream, TryStreamExt};
+use journal_client::broker::{fragments_response, FragmentsRequest, JournalSpec};
+use journal_client::fragments::FragmentIter;
+use journal_client::list::list_journals;
+use journal_client::read::uncommitted::{
+    ExponentialBackoff, JournalRead, ReadStart, ReadUntil, Reader,
+};
+use journal_client::{connect_journal_client, ConnectError};
+use models;
+use schema_inference_autogen::inference_service_server::{
+    InferenceService, InferenceServiceServer,
+};
+use schema_inference_autogen::{InferenceRequest, InferenceResponse};
+
+use futures_util::StreamExt;
+use serde_json::Value;
+use tokio::sync::broadcast::Receiver;
+use tokio::time::sleep;
+use tokio_util::codec::{FramedRead};
+use tokio_util::compat::FuturesAsyncReadCompatExt;
+use tonic::{transport::Server, Request, Response, Status};
+
+pub mod schema_inference_autogen {
+    tonic::include_proto!("schema_inference"); // The string specified here must match the proto package name
+}
+
+#[derive(Debug, Default)]
+pub struct InferenceServiceImpl {
+    pub broker_url: String,
+    pub max_inference_secs: u32
+}
+
+#[tonic::async_trait]
+impl InferenceService for InferenceServiceImpl {
+    // #[tracing::instrument(skip_all, fields(collection=?request.get_ref().flow_collection))]
+    async fn infer_schema(
+        &self,
+        request: Request<InferenceRequest>,
+    ) -> Result<Response<InferenceResponse>, Status> {
+        tracing::debug!("Starting inference");
+
+        let InferenceRequest {
+            flow_collection: collection_name,
+        } = request.get_ref();
+
+        let (abort_tx, abort_rx) = tokio::sync::broadcast::channel::<()>(1);
+
+        let handle = tokio::spawn(async move {
+            sleep(Duration::from_secs(10)).await;
+            abort_tx.send(()).unwrap();
```

Comment:
Doesn't seem like this a safe `unwrap`. Seems like it's possible for all the `abort_rx` receivers to have been dropped prior to this call, for example if they all finish or error out prior to the timeout expiring. It might happen to work, depending on exactly when the `abort_rx` gets dropped, but it seems pretty tricky to rely on that. I think we can just ignore the error here.